### ISSUE TITLE
ParamMap crash

### ISSUE THREAD

**Issue by eldakms** [halide/Halide#2759]
Hi,

I am implementing a simple matrix-vector multiplication and then addition of another vector.
I want to compile the pipeline only once and execute on several threads afterwards.
Unfortunately this causes access violation. When instead I evaluate using std::deferred, all behaves as expected.

OS: Win 10 x64, AVX2, VS2015 Update 3, latest build.

Below is how I do that:

```
std::pair<Halide::Pipeline, Halide::ImageParam> Create(int matrixRowDimension, int matrixColumnDimension)
{
    Halide::ImageParam input(Halide::type_of<float>(), 1);
    Halide::Buffer<float> matrixBuffer(Halide::type_of<float>(), { matrixRowDimension, matrixColumnDimension }, "matrixBuffer");
    Halide::Func matrix;
    Halide::Var x, y;
    matrix(x, y) = matrixBuffer(x, y);

    Halide::Func vec = input;

    Halide::Func prod;
    prod(x, y) = matrix(x, y) * vec(x);

    Halide::Func partial, head, output;
    int vectorizationWidth = 16;
    Halide::RDom k(0, matrixRowDimension / vectorizationWidth);
    Halide::RDom kVector(0, vectorizationWidth);
    Halide::RDom kTail((matrixRowDimension / vectorizationWidth) * vectorizationWidth, matrixRowDimension % vectorizationWidth);

    partial(x, y) = Halide::sum(prod(k* vectorizationWidth + x, y));
    head(y) = Halide::sum(partial(kVector, y));
    output(y) = head(y);

    partial.vectorize(x);
    partial.compute_at(output, y);
    head.compute_at(output, y).vectorize(y, vectorizationWidth, Halide::TailStrategy::GuardWithIf);
    output.compute_root().vectorize(y, vectorizationWidth, Halide::TailStrategy::GuardWithIf);
    output.bound_extent(y, matrixColumnDimension);

    Halide::Buffer<float> anotherVectorBuffer(Halide::type_of<float>(), { matrixColumnDimension }, "vectorBuffer");
    Halide::Func anotherVector, plus;
    anotherVector(x) = anotherVectorBuffer(x);
    plus(x) = anotherVector(x) + output(x);
    plus.compute_root().vectorize(plus.args()[0], 16, Halide::TailStrategy::GuardWithIf);

    auto result = Halide::Pipeline(std::vector<Halide::Func>{ plus });
    Halide::Target t = Halide::get_jit_target_from_environment();
    result.compile_jit(t);
    return std::make_pair(result, input);
}

void main()
{
Halide::Pipeline result;
    Halide::ImageParam p;
    const int InputDimension = 256;
    const int OutputDimension = 9000;

    std::tie(result, p) = Create(InputDimension, OutputDimension);
    std::vector<float> data(InputDimension);
    std::mt19937 r(13);
    for (size_t i = 0; i < InputDimension; ++i)
        data[i] = ((r() % (1 << 16)) / (float)((1 << 16) - 1)) - 0.5f;

    auto work = [InputDimension, OutputDimension](std::vector<float>& data, Halide::Pipeline result, Halide::ImageParam p)
    {
        for (size_t i = 0; i < 50; ++i)
        {
            Halide::Buffer<float> output = Halide::Buffer<float>(OutputDimension);
            Halide::Buffer<float> b = Halide::Buffer<float>(data.data(), std::vector<int> { InputDimension }, "inputBuffer");
            Halide::ParamMap params;
            params.set(p, b);
            result.realize({ output }, Halide::Target(), params);
        }
    };

    const int numberOfAsyncs = 100000;
    std::vector<std::future<void>> results;
    for (size_t i = 0; i < numberOfAsyncs; ++i)
    {
        results.push_back(std::async(std::launch::async, [&data, &result, &work, &p]() { return work(data, result, p); }));
    }

    for (size_t i = 0; i < numberOfAsyncs; ++i)
    {
        results[i].get();
    }
}
```
Any help is appreciated. Thanks!

**Comment by eldakms** [comment]
I have noticed one more thing. If I wrap anotherVectorBuffer into ImageParam instead of using a raw Buffer, I cannot reproduce the crash anymore. Could it be due to aligned/unaligned loads still? (though not clear why this is not reproducible on a single thread).

```
Halide::ImageParam z(Halide::type_of<float>(), 1);
Halide::Buffer<float> anotherVectorBuffer(Halide::type_of<float>(), { matrixColumnDimension }, "vectorBuffer");
z.set(anotherVectorBuffer);
Halide::Func anotherVector, plus;
anotherVector(x) = z(x);
plus(x) = anotherVector(x) + output(x);
plus.compute_root().vectorize(plus.args()[0], 16, Halide::TailStrategy::GuardWithIf);
```
Thanks much!

**Comment by eldakms** [comment]
After more tests, it seems the above does not help.

**Comment by eldakms** [comment]
Any update/confirmation? Thanks!

**Comment by zvookin** [comment]
Not sure why it is crashing exactly, but it has to do with ParamMap not avoiding the `ObjectInstanceRegistry` and lots of calls happening into said registry to register and unregister temporary parameters. (This also seems to happen a lot during compilation.) The registry is thread safe so I don't see why it is failing, but ParamMap should not be making those calls at all so we'll figure out a way to avoid that.

`std::async`, on Mac OS at least, spins up a thread per future. This is really not what anyone wants, ever. There is no way you want 2k threads running Halide code at the same time on a laptop. I've become convinced that most of the thread support in C++11 and such is just completely useless except for toy programs. Likely the answer is we'll have to add a simple way to queue stuff to the Halide thread pool, which given where we're going with optimization and generalization of the model, probably isn't a bad bet.

**Comment by eldakms** [comment]
Than you much for the response!

Yep, I know that functionality of std::async in 11 is pretty limited and in real code we do use a thread pool. The point of this small repro was to stress the system and reveal the race. It is still reproducible on my system even with just 2 std::async calls, but in this case it takes more iterations before the crash. 

This functionality is very important for us, we operate on the server side and cannot afford crashes under load. This is even more important in the jit scenario, because usually you do not get much information at the crash time (i.e. stack traces, etc), so it is very hard to diagnose such kind of issues.

I do not posses enough knowledge about the inners of Halide but have a couple of questions:
1) I would expect that N pipelines do not share anything except for the generated code. Why do they need to share state/do locking? Is creation of a "per-pipeline" ObjectInstanceRegistry too expensive (concurrent stack of object registries or something, so that there is no contention during execution time)? The major concern is whether this locked resource will become a bottleneck in real production system. Is it possible to avoid any sharing of state at realization time between pipelines except for constant/read only state?
2) We do our own thread management, because we know where a particular subgraph will run/how much cache we have/how to split the subgraph in such a way that caches/cores are utilized fully. For us the performance on a multicore systems is really important (both memory moves and compute). So we won't use Halide thread pool and the ParamMap issue is kind of orthogonal to this.
3) What is the current ETA for the possible fix? (I would appreciate if you could also give a reference on the nightly build with the fix, so that I can try it right away)

Thank you very much, really appreciate your help!

**Comment by steven-johnson** [comment]
`std::async` is useless as a portable implementation because of the no-control-over-number-of-threads issue, unfortunately. However, src/ThreadPool.cpp is essentially API-compatible with `std::async` but with strict controls over the threads used.

**Comment by steven-johnson** [comment]
FWIW, I don't think the standard says anything about how many outstanding `std::async` calls can be outstanding at a time; on my Linux box, if I run this as-is, it fails around 32k calls with `terminate called after throwing an instance of 'std::system_error'`; there may well be other issues, but as written, this is gonna be a problem on every system I know of.

**Comment by zvookin** [comment]
Let me try to add a bit of info on why Steven and I are complaining about the `std::async` use here.

It is still not clear why it is crashing. The code using lambdas and async is pretty tricky to figure out and the fact that it is launching thousands or tens of thousands of threads is very unexpected. We have not ruled out that it is running out of resources. Throwing exceptions e.g. due to out of memory, down in the middle of the Halide runtime could easily explain the crash given certain build configurations.

So here we have an issue filed against ParamMap and there is a ton of other stuff going on. There is no way Halide can make code like this C++ stuff reliable. Yeah there might be a bug in the ParamMap support but two of us are a few hours in and all we have is an efficiency issue on the object registry. We'll fix that, and perhaps find the cause of the crash, but the bug report has cost more time and effort than is desired.

**Comment by eldakms** [comment]
I will reproduce this tomorrow with just a few outstanding async calls. Sorry different timezone.

Sent from my Windows 10 phone

**Comment by steven-johnson** [comment]
FYI, I'm pretty sure that https://github.com/halide/Halide/issues/2765 is also happening here; not sure if this is the only cause here or not though.

**Comment by zvookin** [comment]
I have a fix for the user context thread safety issue, which may be the root cause of the crash. Between that and Steven's registry fix, I think we may have taken care of everything wrong with Halide here.

Mac OS X has a hard limit on the number of threads that can be created within a single process. `std::async` just creates a thread for every future. The combination of these results in the test program hanging, if the async requests take long enough to allow the main thread to create too many threads. (That it hangs is problematic, but probably has nothing to do with Halide other than that we may be compiling with exception handling turned off.)

**Comment by eldakms** [comment]
Great, thanks a lot! Could you please point me to the binaries when the fix will be available? Thanks!

**Comment by eldakms** [comment]
Btw, here is the repro with just two threads if you still need it (I changed to std::thread, the rest is the same). This crashes on Windows on the current binaries:

```
std::pair<Halide::Pipeline, Halide::ImageParam> create(int matrixRowDimension, int matrixColumnDimension)
{
    Halide::ImageParam input(Halide::type_of<float>(), 1);
    Halide::Buffer<float> matrixBuffer(Halide::type_of<float>(), { matrixRowDimension, matrixColumnDimension }, "matrixBuffer");
    Halide::Func matrix;
    Halide::Var x, y;
    matrix(x, y) = matrixBuffer(x, y);

    Halide::Func vec = input;

    Halide::Func prod;
    prod(x, y) = matrix(x, y) * vec(x);

    Halide::Func partial, head, output;
    int vectorizationWidth = 16;
    Halide::RDom k(0, matrixRowDimension / vectorizationWidth);
    Halide::RDom kVector(0, vectorizationWidth);
    Halide::RDom kTail((matrixRowDimension / vectorizationWidth) * vectorizationWidth, matrixRowDimension % vectorizationWidth);

    partial(x, y) = Halide::sum(prod(k* vectorizationWidth + x, y));
    head(y) = Halide::sum(partial(kVector, y));
    output(y) = head(y);

    partial.vectorize(x);
    partial.compute_at(output, y);
    head.compute_at(output, y).vectorize(y, vectorizationWidth, Halide::TailStrategy::GuardWithIf);
    output.compute_root().vectorize(y, vectorizationWidth, Halide::TailStrategy::GuardWithIf);
    output.bound_extent(y, matrixColumnDimension);

    Halide::Buffer<float> anotherVectorBuffer(Halide::type_of<float>(), { matrixColumnDimension }, "vectorBuffer");
    Halide::Func anotherVector, plus;
    anotherVector(x) = anotherVectorBuffer(x);
    plus(x) = anotherVector(x) + output(x);
    plus.compute_root().vectorize(plus.args()[0], 16, Halide::TailStrategy::GuardWithIf);

    auto result = Halide::Pipeline(std::vector<Halide::Func>{ plus });
    Halide::Target t = Halide::get_jit_target_from_environment();
    result.compile_jit(t);
    return std::make_pair(result, input);
}

const int InputDimension = 256;
const int OutputDimension = 9000;
const int NumberOfRealizationsPerThread = 100000;

void shared_realize(Halide::Pipeline pipeline, Halide::ImageParam p)
{
    std::vector<float> data(InputDimension, 0);
    for (size_t i = 0; i < NumberOfRealizationsPerThread; ++i)
    {
        Halide::Buffer<float> output = Halide::Buffer<float>(OutputDimension);
        Halide::Buffer<float> b = Halide::Buffer<float>(data.data(), std::vector<int> { InputDimension }, "inputBuffer");
        Halide::ParamMap params;
        params.set(p, b);
        pipeline.realize({ output }, Halide::Target(), params);
    }
}

void main()
{
    Halide::ImageParam p;
    Halide::Pipeline pipeline;
    std::tie(pipeline, p) = create(InputDimension, OutputDimension);

    std::thread t[2];
    for (int i = 0; i < 2; ++i)
    {
        t[i] = std::thread(shared_realize, pipeline, p);
    }

    for (int i = 0; i < 2; ++i)
    {
        t[i].join();
    }
}
```

**Comment by eldakms** [comment]
I tried the latest master, cannot reproduce, closing.
